lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -392,14 +467,37 @@ def _gen_month_range(self, index: str, start: datetime, end: datetime) -> str:
             )
         return indices
 
-    def assemble(self, json_data: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+    def preprocess(self, client_json: Dict[AnyStr, Any]) -> Any:
+        """
+        Given the client Request payload, perform any preprocessing activities
+        necessary prior to constructing an Elasticsearch query.
+
+        The base class assumes no preprocessing is necessary, and returns True
+        to indicate that the Elasticsearch operation should continue; this can
+        be overridden by subclasses.
+
+        Args:
+            client_json: Request JSON payload
+
+        Raises:
+            Any errors in the postprocess method shall be reported by
+            exceptions which will be logged and will terminate the operation.
+
+        Returns:
+            None if Elasticsearch query shouldn't proceed, or context to pass
+            to the assemble and postprocess methods.
+        """
+        return True
+
+    def assemble(self, json_data: Dict[AnyStr, Any], context: Any) -> Dict[AnyStr, Any]:
         """
-        assemble Assemble the Elasticsearch parameters
+        Assemble the Elasticsearch parameters.
 
         This must be overridden by the subclasses!
 
         Args:
             json_data: Input JSON payload, processed by type conversion
+            context: Value returned by preprocess method
```

Comment:
The point is that the `post` method that calls these follows a sequence, and the subclass `assemble` and `postprocess` __will__ always be called with the value returned by `preprocess`. I don't see any point in saying anything less specific here.

The only way that would change is if the base class is changed (requiring documentation updates) or if we build a "sub baseclass" for a new subset that overrides these abstract methods, again requiring new documentation.

Reply:
I guess I was aiming for having the documentation provide something more than a description of the mechanism (which really has no bearing on what this function does).  I think the point of the context is to allow the preprocess method and the subsequent two calls to communicate and coordinate in a modular fashion.  Thus, the interest for this function is not so much in how the context was created (to wit, the fact that the preprocess method _returned it_) as it is in how the contents of it can be used.